File: apps/rust-test-server/src/graphql/resolvers/media.rs
/*!
# Media Resolvers

This module handles media upload and processing GraphQL operations including:
- Media upload functionality
- Batch media uploads
- Image processing and editing
- Media variants and optimization

Based on TypeScript implementation in apps/server/src/graphql/resolvers/media.ts
*/

use crate::graphql::{
    GraphQLEngine, GraphQLValue, GraphQLError, FieldContext
};
use crate::services::database_simple::DatabaseService;
use std::sync::Arc;
use serde_json::{json, Value};

/// Add media resolvers to the GraphQL engine
pub fn add_media_resolvers(engine: &mut GraphQLEngine, _db_service: Arc<DatabaseService>) {
    // TODO: Implement media resolvers
    // Mutation resolvers:
    // - uploadMedia(input)
    // - uploadMediaBatch(input)
    // - processEditedImage(input)
}

/// Upload media files
pub fn upload_media_resolver(ctx: &FieldContext, _db_service: &DatabaseService) -> Result<GraphQLValue, GraphQLError> {
    let input = ctx.field_info.arguments
        .get("input")
        .ok_or_else(|| GraphQLError::new("Missing required argument 'input'"))?;

    // TODO: Implement actual media upload
    // - Generate upload URLs for Cloudflare Images
    // - Create media records in database
    // - Handle image optimization and variants
    
    let count = input.get("count")
        .and_then(|v| v.as_i64())
        .unwrap_or(1);

    let mut uploads = Vec::new();
    for i in 0..count {
        uploads.push(json!({
            "uploadURL": format!("https://upload.imagedelivery.net/account/upload_{}", i),
            "id": format!("media_{}", i)
        }));
    }

    let response = json!({
        "uploads": uploads
    });

    Ok(GraphQLValue::from(response))
}

/// Upload batch of media files
pub fn upload_media_batch_resolver(ctx: &FieldContext, _db_service: &DatabaseService) -> Result<GraphQLValue, GraphQLError> {
    let input = ctx.field_info.arguments
        .get("input")
        .ok_or_else(|| GraphQLError::new("Missing required argument 'input'"))?;

    // TODO: Implement actual batch media upload
    // - Process array of media items
    // - Create media records with proper ordering
    // - Associate with post or draft post
    
    let media_items = input.get("mediaItems")
        .and_then(|v| v.as_array())
        .ok_or_else(|| GraphQLError::new("Missing or invalid mediaItems in input"))?;

    let mut media = Vec::new();
    for (index, item) in media_items.iter().enumerate() {
        let image_id = item.get("imageId")
            .and_then(|v| v.as_str())
            .ok_or_else(|| GraphQLError::new("Missing imageId in media item"))?;

        let order = item.get("order")
            .and_then(|v| v.as_i64())
            .unwrap_or(index as i64);

        media.push(json!({
            "id": format!("media_{}", index),
            "type": "image",
            "storageKey": image_id,
            "order": order,
            "createdAt": "2024-01-01T00:00:00Z"
        }));
    }

    let response = json!({
        "media": media
    });

    Ok(GraphQLValue::from(response))
}

/// Process edited image with filters and adjustments
pub fn process_edited_image_resolver(ctx: &FieldContext, _db_service: &DatabaseService) -> Result<GraphQLValue, GraphQLError> {
    let input = ctx.field_info.arguments
        .get("input")
        .ok_or_else(|| GraphQLError::new("Missing required argument 'input'"))?;

    let original_image_id = input.get("originalImageId")
        .and_then(|v| v.as_str())
        .ok_or_else(|| GraphQLError::new("Missing originalImageId in input"))?;

    let _editing_metadata = input.get("editingMetadata")
        .ok_or_else(|| GraphQLError::new("Missing editingMetadata in input"))?;

    // TODO: Implement actual image processing
    // - Apply filters, adjustments, and crops
    // - Generate processed image variants
    // - Store editing metadata
    // - Return processed image information
    
    let response = json!({
        "processedImageId": format!("processed_{}", original_image_id),
        "variants": [
            "thumbnail",
            "medium",
            "large"
        ],
        "originalImageId": original_image_id
    });

    Ok(GraphQLValue::from(response))
}

// Helper functions for image processing

/// Generate image variants for different use cases
fn generate_image_variants(_image_id: &str) -> Vec<String> {
    // TODO: Implement actual variant generation
    vec![
        "thumbnail".to_string(),
        "small".to_string(),
        "medium".to_string(),
        "large".to_string(),
        "original".to_string(),
    ]
}

/// Apply photo filters to image
fn apply_photo_filters(_image_id: &str, _filters: &[Value]) -> Result<String, GraphQLError> {
    // TODO: Implement actual filter application
    Ok("filtered_image_id".to_string())
}

/// Apply photo adjustments to image
fn apply_photo_adjustments(_image_id: &str, _adjustments: &Value) -> Result<String, GraphQLError> {
    // TODO: Implement actual adjustment application
    Ok("adjusted_image_id".to_string())
}

/// Apply crop data to image
fn apply_crop_data(_image_id: &str, _crops: &[Value]) -> Result<String, GraphQLError> {
    // TODO: Implement actual cropping
    Ok("cropped_image_id".to_string())
}

/// Validate editing metadata structure
fn validate_editing_metadata(metadata: &Value) -> Result<(), GraphQLError> {
    // TODO: Implement comprehensive validation
    // - Check required fields
    // - Validate value ranges
    // - Ensure crop dimensions are valid
    
    if metadata.get("filters").is_none() && 
       metadata.get("adjustments").is_none() && 
       metadata.get("crops").is_none() {
        return Err(GraphQLError::new("At least one editing operation required"));
    }
    
    Ok(())
}

File: apps/server/src/graphql/resolvers/media.ts
import { createImageService } from '../../lib/images/imageService';
import type { ContextType } from '../../types';

export const mediaResolvers = {
  Mutation: {
    async uploadMedia(
      _parent: unknown,
      { input }: { input: { count?: number } },
      context: ContextType,
    ) {
      if (!context?.user) {
        throw new Error('Authentication required');
      }

      try {
        const imageService = createImageService(context.env);
        const count = input.count || 1;

        if (count < 1 || count > 10) {
          throw new Error('Count must be between 1 and 10');
        }

        if (count === 1) {
          const upload = await imageService.getUploadUrl(context.user.id);
          return { uploads: [upload] };
        }

        const uploads = await imageService.getBatchUploadUrls(context.user.id, 'user', count);
        return { uploads };
      } catch (error) {
        console.error('Upload URL generation error:', error);
        throw new Error(
          `Failed to generate upload URLs: ${error instanceof Error ? error.message : 'Unknown error'}`,
        );
      }
    },

    async uploadMediaBatch(
      _parent: unknown,
      { input }: { input: { mediaItems: any[] } },
      context: ContextType,
    ) {
      if (!context?.user) {
        throw new Error('Authentication required');
      }

      try {
        const imageService = createImageService(context.env);
        const { mediaItems } = input;

        if (!mediaItems || !Array.isArray(mediaItems) || mediaItems.length === 0) {
          throw new Error('Media items array is required');
        }

        if (mediaItems.length > 10) {
          throw new Error('Maximum 10 media items per batch');
        }

        // Validate all items have required fields
        for (const item of mediaItems) {
          if (!item.imageId) {
            throw new Error('All items must have imageId');
          }
        }

        // Add userId to all items
        const enrichedItems = mediaItems.map((item, index) => ({
          ...item,
          userId: context.user.id,
          order: item.order ?? index,
        }));

        const createdMedia = await imageService.createBatch(enrichedItems);
        return { media: createdMedia };
      } catch (error) {
        console.error('Batch media creation error:', error);
        throw new Error(
          `Failed to create batch media: ${error instanceof Error ? error.message : 'Unknown error'}`,
        );
      }
    },

    async processEditedImage(
      _parent: unknown,
      { input }: { input: { originalImageId: string; editingMetadata: any } },
      context: ContextType,
    ) {
      if (!context?.user) {
        throw new Error('Authentication required');
      }

      try {
        const imageService = createImageService(context.env);
        const { originalImageId, editingMetadata } = input;

        if (!originalImageId || !editingMetadata) {
          throw new Error('Original image ID and editing metadata are required');
        }

        const result = await imageService.processEditedImage({
          originalImageId,
          editingMetadata,
          userId: context.user.id,
        });

        // Optimize the processed image for different variants
        await imageService.optimizeForVariants(result.processedImageId, result.variants);

        return {
          processedImageId: result.processedImageId,
          variants: result.variants,
          originalImageId,
        };
      } catch (error) {
        console.error('Image processing error:', error);
        throw new Error(
          `Failed to process edited image: ${error instanceof Error ? error.message : 'Unknown error'}`,
        );
      }
    },
  },
};
